network_simulator/flow.py

```diff
@@ -125,38 +173,428 @@ def handle_packet_loss(self, packet_id, loss_type, main_event_loop):
         :param MainEventLoop main_event_loop: main Event loop for further
         Event scheduling.
         """
-        pass
+        logger.debug("Flow %s lost packet %d due via PacketLossType %d",
+                     self.flow_id, packet_id, loss_type)
 
     @abstractmethod
-    def handle_packet_success(self, packet_id):
+    def handle_packet_success(self, packet, statistics, curr_time):
         """
         A handler for dealing with packets that successfully completed a
         round trip. This will adjust window sizes and make other state
-        changes, depending on the TCP congestion algorithm.
-        :param packet_id: ID of packet that completed a round trip
+        changes, depending on the TCP congestion algorithm. It will also
+        update common metadata and statistics.
+
+        The sending of additional Packets (e.g. if the window size is grown)
+        must be handled elsewhere, since this can be subclass-dependent.
+        :param Packet packet: The AckPacket that completed a round trip.
+        :param Statistics statistics: The Statistics to update.
+        :param float curr_time: The current simulation time (in seconds).
         """
-        pass
+        assert isinstance(packet, AckPacket)
+
+        # Update RTTs
+        sent_time = packet.data_packet_start_time_sec
+        curr_rtt = curr_time - sent_time
+        self.packet_rtts.append((packet.packet_id, curr_rtt))
+
+        # All Flows need to properly update state variables related to
+        # transiting/retransmitted packets.
+        self.packets_in_transit.remove(packet.packet_id)
+        if packet.packet_id in self.gap_retrans_packets:
+            self.gap_retrans_packets.remove(packet.packet_id)
+
+        # No longer need to do anything on timeouts for this Packet.
+        del self.num_timeouts_pending[packet.packet_id]
+
+        # All Flows should record these statistics.
+        statistics.flow_packet_received(flow=self, ack_packet=packet,
+                                        curr_time=curr_time)
+
+    def retransmit_if_possible(self, packet_id, main_event_loop):
+        """
+        Attempts to retransmit the given Packet by scheduling a
+        FlowSendPacketsEvent. This will only be done, however, when there's
+        enough space in the window. Otherwise, the ID is added to a buffer.
+        :return True if we were able to immediately schedule a retransmit.
+        """
+        # TODO(laksh): Could add ID to gap_retrans_packets *here*, if this
+        # retransmission is due to a gap. Might be cleaner.
+
+        # Get size of packets_in_transit if we were to add this Packet.
+        packets_in_transit_copy = copy.copy(self.packets_in_transit)
+        packets_in_transit_copy.add(packet_id)
+
+        if len(packets_in_transit_copy) > int(self.window_size_packets):
```

Comment:
This is intended. The max window size needs to be a float, or else the algorithms to update it will not work. But when using it for filling an "entire window" of packets, we need to round it down. Example: a max window size of 6.34 packets means at most 6 packets should be in transit at once.